Enable static assets in Learning Core XBlock Runtime

ormsbee opened the pull request:
## Description

This enable the viewing of static assets in the Learning Core XBlock Runtime. It is version aware, so
calling the XBlock iframe with different versions will properly reflect the right assets for that
version of the Component.

Implements https://github.com/openedx/edx-platform/issues/35659

It does not currently support adding assets, and that must still be done via the `add_assets_to_component`
management command.

Copy-paste functionality will be a separate PR.

ormsbee committed:
feat: versioned asset support for Learning Core XBlock runtime

Add support for displaying static assets in the Learing Core XBlock
runtime via "/static/asset-name" style substitutions in the OLX. This is
currently used for new Content Library components.

Static asset display is version-aware, so viewing older versions of the
XBlock content via the embed view will show the appropriate assets for
that version.

bradenmacdonald commented on openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py:
```diff
@@ -300,19 +337,90 @@ def _get_component_from_usage_key(self, usage_key):
+        usage_key = block.usage_key
+        component = self._get_component_from_usage_key(usage_key)
+
+        block_version = block._runtime_requested_version  # pylint: disable=protected-access
+        if block_version == LatestVersion.DRAFT:
+            component_version = component.versioning.draft
+        elif block_version == LatestVersion.PUBLISHED:
+            component_version = component.versioning.published
+        else:
+            component_version = component.versioning.version_num(block_version)
+
+        return component_version
```
Do you think it makes sense to combine this with the similar logic in `get_block()`?

https://github.com/openedx/edx-platform/blob/ecbedbcf78cd74c907f18d76b77a64e31ccf78fe/openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py#L183-L193

ormsbee replied:
I'm kind of on the fence about it. When I took a brief sketch at extracting out the common part here, the helper naming and indirection just seemed to add more confusion than it was worth. I don't feel strongly about it, but I'm inclined to just leave the bit of redundancy.

bradenmacdonald replied:
Fine with me!

bradenmacdonald commented on openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py:
```diff
@@ -300,19 +337,90 @@ def _get_component_from_usage_key(self, usage_key):
+        then this method will be called with asset_path="test.png" and should
+        return a URL like:
+
+          http://studio.local.edly.io:8001/library_assets/cd31871e-a342-4c3f-ba2f-a661bf630996/static/test.png
```
Nit: we're moving toward `local.openedx.io` so I'd recommend just using that in all docs going forward. e.g. the new [`npm run dev` command](https://github.com/openedx/frontend-app-authoring?tab=readme-ov-file#startup) for running the MFEs in dev mode on your host (in Authoring & Learning MFEs only) uses that domain exclusively.

bradenmacdonald commented on openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py:
```diff
@@ -300,19 +337,90 @@ def _get_component_from_usage_key(self, usage_key):
+        Implementation Details
+        ----------------------
+
+        The standard XBlock "static/{asset_path}"" substitution strips out the
```
```suggestion
        The standard XBlock "static/{asset_path}" substitution strips out the
```

bradenmacdonald commented on openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py:
```diff
@@ -300,19 +337,90 @@ def _get_component_from_usage_key(self, usage_key):
+        On the backend, we only store the asset once.
+
+        TODO: Like get_block, we currently assume that we're using the Draft
+        version. This should be a runtime parameter.
```
```suggestion
```
Seems like this is already resolved.

bradenmacdonald commented on openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py:
```diff
@@ -230,6 +233,40 @@ def get_block(self, usage_key, for_parent=None, *, version: int | LatestVersion
+        return [
+            StaticFile(
+                name=cvc.key,
+                url=reverse(
```
Should this also be absolute, like `_lookup_asset_url` returns?

ormsbee replied:
Probably. I'll just do that and see if anything breaks. 😛

bradenmacdonald commented on openedx/core/lib/xblock_serializer/block_serializer.py:
```diff
@@ -37,6 +37,20 @@ def __init__(self, block):
+            self.static_files.extend(
+                block.runtime.get_block_assets(block)
+            )
+            return
```
Nit: I'd recommend moving this code and the code below into a new `_find_static_assets` method or something like that. Or use an explicit `else:` case. I'm just worried  that someone might not notice the conditional early `return`  here, and add some unrelated code to the end of the constructor, and later encounter a bug when it sometimes doesn't run.

bradenmacdonald reviewed (commented):
Seems to be working well with the test content I have :)

bradenmacdonald commented on openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py:
```diff
@@ -360,7 +365,7 @@ def _lookup_asset_url(self, block: XBlock, asset_path: str) -> str | None:
         return a URL like:
 
-          http://studio.local.edly.io:8001/library_assets/cd31871e-a342-4c3f-ba2f-a661bf630996/static/test.png
+          http://local.openedx.io:8001/library_assets/cd31871e-a342-4c3f-ba2f-a661bf630996/static/test.png
```
Err sorry that should be `http://studio.local.openedx.io:8001/` or `http://local.openedx.io:8000/`

ormsbee replied:
Right. I should know better. Is late, brain no-workie good. <sigh>

bradenmacdonald reviewed (approved):
Looks great! I tested viewing assets in the new library Preview UI in the authoring MFE and it's all working as expected.

![Screenshot](https://github.com/user-attachments/assets/8c62fdd8-b644-4608-8a29-dc1a5f3ab9a2)

ormsbee committed:
test: add data.py to acceptable isolated app imports

Per OEP-49, both api.py and data.py are allowed to be imported into
other apps:

https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0049-django-app-patterns.html#api-py
